Creating a settings object that is aware of the microsite settings.

felipemontoya opened the pull request:
This PR creates a settings object which can look for a setting in the microsite configuration and default to the regular django.conf settings value when the microsite does not have a particular override.

This should not have any impact on running code, but will allow a developer to call

    from microsite_configuration import settings
instead of 

    from django.conf import settings

and then use the settings object as it would normally with django. e.g.

    settings.SITE_NAME

The current instances of microsite.get_value are not being changed in this PR.

This is a part of the changes contained in #7867 . I'm spliting that PR to have more isolated changes and make it easier to merge.

edx-webhook commented:
Thanks for the pull request, @felipemontoya! I've created [OSPR-605](https://openedx.atlassian.net/browse/OSPR-605) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams. 

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ('this must be merged by XX date', and why that is)
- partner information ('this is a course on edx.org')
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the Github pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

sarina commented:
@felipemontoya how does this relate to https://github.com/edx/edx-platform/pull/7867?

sarina commented:
@felipemontoya ignore my last comment, reading comprehension fail.

Is @chrisndodge going to take a shot at reviewing this?

felipemontoya commented:
@sarina, I'm pinging him now. 

@chrisndodge, I opened this, just to discuss the MicrositeAwareSettings object.

FYI, @benpatterson @cahrens @andy-armstrong

cahrens commented on common/djangoapps/microsite_configuration/__init__.py:
```diff
@@ -1 +1,19 @@
+    from microsite_configuration import settings
+    """
+
+    def __getattr__(self, name):
```
Would it make sense to write a unit test for this?

chrisndodge commented on common/djangoapps/microsite_configuration/__init__.py:
```diff
@@ -1 +1,19 @@
+    """
+
+    def __getattr__(self, name):
+        return microsite.get_value(name, base_settings.__getattr__(name))
```
One thing all users of microsites will have to do is to - for the short term - provide duplicate definitions inside of MICROSITE_CONFIGURATION for "platform_name" == "PLATFORM_NAME". I don't think it should be within the scope of these changes to touch all the points where we have to change the capitalization.

We'll have to change our configuration on our sites at edx.org, so this PR will have to be coordinated with some DevOps changes.

Either that or have some ugly logic somewhere (say in get_value()) which treats "platform_name" as "PLATFORM_NAME".

Not sure which is better/worse.

felipemontoya commented:
I have two concerns about this object. 

The first is that the microsite object defined at the settings

    MICROSITE_CONFIGURATION = {
        "devsite": {
            "domain_prefix": "devsite",
            "university": "devsite",
            "platform_name": "Development microsite",
            ...
        }
    }

would have to map the current format of the settings. For example ENABLE_MKTG_SITE would need to be inside a FEATURES dict, or it wont find it. Optionally we can add a transation layer here.

The second concern is that this new settings object would override any setting in the django settings object in the scope of the file where the import header was changed. Optionally we could whitelist the settings that are to be looked for in the microsite and those directly go to the django settings.

chrisndodge commented:
@felipemontoya, right, I think - in essence - we'll need to update our configuration so that MICROSITE_CONFIGURATION has the same structure as 'settings.py', e.g. also have a FEATURES subdict.

There are a few other known users of microsites. @natea is one who - AFAIK - uses Microsites a lot. Seems good to bring him into the discussion, particularly if we are going to force configuration changes. I know we have a few other partners that use microsites, XeutangX is claiming to use microsites on the edx-platform Wiki. I'm trying to get technical for them.

Seems like it make sense to post something to the Open edX mailing list to see who else might need to be consulted. What do you think?

nedbat commented:
I'm wondering if we can do this in a way that people don't have to consciously use "microsites", but instead can just be using "openedx.settings", or something.

felipemontoya commented:
@chrisndodge any updates on this? I wrote to the edx-code list more than a week ago, and nobody replied or asked further questions. I don't know if it means, no more people is using microsites.

@nedbat At first I thought it was simple, just move the location of the settings declaration to any file under openedx/core/djangoapps or even to openedx/init.py. Now Im looking at it again, and the thing is this does require the 

    from microsite_configuration import microsite

line, which uses a package defined at the lms. Would that be a problem for studio?

sarina commented:
@felipemontoya [OSPR-605](https://openedx.atlassian.net/browse/OSPR-605) should be in @chrisndodge 's team backlog - it will be prioritized into a sprint for review.

@chrisndodge can you confirm that that ticket shows up in your backlog? I don't know what board your team uses for this type of work.

sarina commented:
@chrisndodge could you please work to get this prioritized into a sprint? @felipemontoya has been waiting for a review for over six weeks.

@felipemontoya - this will need a rebase atop master to get a fresh build, as many changes have been made. It would also be good if you could make sure you get a clean build.

felipemontoya commented:
@sarina, I just rebased it and had the test run locally, which they did. The build should be clean now.

@nedbat I added a declaration of the microsite aware settings object to openedx.conf. Now this can be imported as:

    from openedx.conf import settings

I tried to do it in the openedx.core package, but it gave me errors on common/lib/xmodule

@chrisndodge I'm still lagging behind the work on the selectable backend to store the microsites. I'll try to get that running and in a PR soon.

chrisndodge commented:
Steve/part ha, Do we have a story already for this ospr review?

Sent from my iPhone

> On Jul 14, 2015, at 10:12 AM, Sarina Canelake <notifications@github.com> wrote:
> 
> @chrisndodge could you please work to get this prioritized into a sprint? @felipemontoya has been waiting for a review for over six weeks.
> 
> @felipemontoya - this will need a rebase atop master to get a fresh build, as many changes have been made. It would also be good if you could make sure you get a clean build.
> 
> —
> Reply to this email directly or view it on GitHub.
>

sarina commented:
@chrisndodge yes, [OSPR-605](https://openedx.atlassian.net/browse/OSPR-605) should be in your team's backlog. I have verified I can see OSPR tickets in the Solutions team backlog.

felipemontoya commented:
@sarina I though that quality was also running locally. Since it wasnt, I'll address the quality issues and push them later along with the feedback of the review to avoid running the tests again unnecesarily

chrisndodge commented:
Hi Felipe,

Sorry for the long radio silence. I'm in the middle of a deadline right now, so my first available window to follow-up on this PR is in early august.

I believe the main topic to discuss is backward compatibility regarding the configuration schema. Just pinging @natea again to see if we can bring him into the conversation.

Thank you for your patience and perserverence,

Chris

natea commented:
I would have loved to have this PR land in Cypress, but it's unlikely
that's going to happen given Chris' schedule. Like Chris, I've been chasing
deadlines and haven't been able to give this PR the attention it deserves!
Let me know what I can do to help accelerate the review process.

On Wed, Jul 15, 2015 at 3:11 PM, chrisndodge <notifications@github.com>
wrote:

> Hi Felipe,
>
> Sorry for the long radio silence. I'm in the middle of a deadline right
> now, so my first available window to follow-up on this PR is in early
> august.
>
> I believe the main topic to discuss is backward compatibility regarding
> the configuration schema. Just pinging @natea <https://github.com/natea>
> again to see if we can bring him into the conversation.
>
> Thank you for your patience and perserverence,
>
> Chris
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/8153#issuecomment-121717673>.
>

sarina commented:
You guys have definitely already missed the Cypress train; we're basing it off of the 7/6 release and are only accepting critical bug and performance fixes into it at this point in time. For your reference we aim to cut Dogwood in late October for a late November release.

natea commented:
What's the ETA for a Cypress RC? We're awaiting with eager anticipation!
Good to know the rough dates for Dogwood release. thanks.

On Wed, Jul 15, 2015 at 3:17 PM, Sarina Canelake <notifications@github.com>
wrote:

> You guys have definitely already missed the Cypress train; we're basing it
> off of the 7/6 release and are only accepting critical bug and performance
> fixes into it at this point in time. For your reference we aim to cut
> Dogwood in late October for a late November release.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/8153#issuecomment-121718975>.
>

sarina commented:
Nate - that's a good question; we've found a fair number of bugs related to
bringing up Cypress and are working those through. I recommend reaching out
on the mailing list and asking; I'll prod DB into answering :)

On Wed, Jul 15, 2015 at 3:38 PM, Nate Aune <notifications@github.com> wrote:

> What's the ETA for a Cypress RC? We're awaiting with eager anticipation!
> Good to know the rough dates for Dogwood release. thanks.
>
> On Wed, Jul 15, 2015 at 3:17 PM, Sarina Canelake <notifications@github.com
> >
> wrote:
>
> > You guys have definitely already missed the Cypress train; we're basing
> it
> > off of the 7/6 release and are only accepting critical bug and
> performance
> > fixes into it at this point in time. For your reference we aim to cut
> > Dogwood in late October for a late November release.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/edx/edx-platform/pull/8153#issuecomment-121718975>.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/8153#issuecomment-121723159>.
>

felipemontoya commented:
@chrisndodge any updates on this? I would also welcome any feedback on PR #8971 if you have a chance to look at it.

chrisndodge commented:
Hi Felipe, sorry for my lack of attentiveness - I'm under a deadline now. Is this the only changeset you want to merge now?

felipemontoya commented:
I finally split the first PR into two, this one and #8971 that handles the selectable backend interface to store the microsites. That one is not yet ready to merge, but I wanted to hear some initiall feedback before commiting more time to it.

chrisndodge commented:
Hi Felipe,

This looks fine and reflects what we discussed. The bigger is obviously what is contained in https://github.com/edx/edx-platform/pull/8971

We also need to discuss if we change some of the LMS call sites. To really take advantage of this change, we'd need to change how our configuration is organized (since we'd still be using what is on disk). For example what used to be

```
microsite.get_value('ENABLE_FOO', <default-which-is-normally-django-settings>)
```

should become

```
microsite.conf.FEATURES['ENABLE_FOO']
```

So then what was in configuration as:

```
MICROSITE_CONFIGURATION: {
    'my-site': {
         'ENABLE_FOO': True
    }
}
```

should become:

```
MICROSITE_CONFIGURATION: {
    'my-site': {
        'FEATURES': {
             'ENABLE_FOO': True
        }
    }
}
```

However, as we discussed, this would be a schema change and edX would have to change our configuration. I'd imagine we'd be willing to do so as we have a limited number of microsites now, so such a change might be best to be done sooner rather than later. But I'll have to swing around and talk to some folks.

felipemontoya commented:
Hi Chris, 
I thought about this for a while and here are my conclusions:

    microsite.conf.FEATURES['ENABLE_FOO']

The line should be `settings.FEATURES['ENABLE_FOO']` where settings is defined as `from microsite_configuration import settings` or `from openedx.conf import settings`

Now, more important is that calls to 'FEATURES' actually could not always work, since you would get a complete but different object which may not include what is defined in the regular settings.

Lets say you have the following configurations:


    FEATURES['ENABLE_FOO'] = True
    FEATURES['ENABLE_BAR'] = True

    MICROSITE_CONFIGURATION: {
        'my-site': {
            'FEATURES': {
                 'ENABLE_FOO': False
            }
        }
    }

and you call it using the openedx.conf notation

    from openedx.conf import settings

    if settings.FEATURES['ENABLE_BAR']:
        pass

Now when you call this from a not microsited domain, you get `True` because the 
FEATURES dict contains the flag. But if you call it from `my-site` now you get
a key error, for the FEATURES dict only contains the ENABLE_FOO flag.

For this to work, a merge of both dictionaries is required. I tried it locally and I am adding the merge method since it think it is a better alternative than to be required to define all the list of FEATURES in a microsite, only for the ability to modify one of them. Any thoughts on this?

chrisndodge commented on common/djangoapps/microsite_configuration/__init__.py:
```diff
@@ -1 +1,48 @@
+            temp.update(microsite.get_value(name, {}))
+            return temp
+        else:
+            return base_settings.__getattr__(name)
```
I wonder if there is a way to cache this merging of dictionaries, at least on a per-request basis. It seems like this merging could get called alot over the lifetime of the request processing.

Could we put a cache on the thread_local definition that is in middleware.py?

felipemontoya replied:
Per process lifetime would complicate things for microsites defined in the db, because then a restart of the process would be required if any change happened in the db (and that would make no sense), so I lean towards the cache.

I think a get_cached_value method could be exposed on the microsite module and the merge_dict function would use that.

chrisndodge replied:
Ah yea, gotcha. Forgot that you ultimately want to move to using a database source configuration backend. Yea, that would not work with process lifetime caching.

I agree that a thread_local cache that is scoped to the lifetime of the HTTP request would be a good thing.

sarina commented:
@chrisndodge what's going on with this pull request?

chrisndodge commented:
I made some suggestions to @felipemontoya about caching the settings lookups. Waiting to hear back.

felipemontoya commented:
I have this on my todo list, I'm just super busy with a cypress migration today. But I will give it a bump before the week is over.

sarina commented:
Chris - if you move the OSPR ticket to "waiting on author" when you're
awaiting a response, and "Engineering Review" when you're actively
reviewing, it helps me understand where the review is at (and I bug you
less).

On Wed, Aug 26, 2015 at 1:13 PM, Felipe Montoya <notifications@github.com>
wrote:

> I have this on my todo list, I'm just super busy with a cypress migration
> today. But I will give it a bump before the week is over.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/8153#issuecomment-135111356>.
>

felipemontoya commented:
@chrisndodge, I changed the dictionary merge to be part of the functions defined in the microsite module, since it is the one with access to the threading local data.

I also changed the backends definitions on PR #8971 to match the required changes here.

sarina commented:
@felipemontoya you have a small number of quality errors and one failing test that's probably worth looking into.

felipemontoya commented:
Hi @sarina 
The quality errors, are easy, I'll address them. If a rebase is needed I might as well do it on the rebase commit.

The failing test is a puzzle to me. Locally it is ok, and from jenkins I get

    DEBUG: Using MongoDB for static content serving at host=localhost port=27017 db=xcontent\nutil.models: ERROR: String decompression failed. There may be corrupted data in the database: 

Which I don't know what to make of since I have no control of what is in the database in that test.

Any ideas that could help me here?

sarina commented:
Hmm yeah that looks infrastructure related - I'll ask someone on test.

sarina commented:
@felipemontoya the branch for the PR needs to be rebased to get rid of the string compression failure msg. fixed in this PR https://github.com/edx/edx-platform/pull/9328

felipemontoya commented:
@sarina I rebased the PR on to master, and it fixed the python test indeed. Now it fails on a random js test.
Nothing in this PR alters the behavior of the Instructor page or js at all. Could you help me on how to proceed?

sarina commented:
I think that's still test troubles. Try rebasing again in a few days.

On Fri, Aug 28, 2015 at 1:50 PM, Felipe Montoya <notifications@github.com>
wrote:

> @sarina <https://github.com/sarina> I rebased the PR on to master, and it
> fixed the python test indeed. Now it fails on a random js test.
> Nothing in this PR alters the behavior of the Instructor page or js at
> all. Could you help me on how to proceed?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/8153#issuecomment-135843999>.
>

felipemontoya commented:
Hi @sarina, I rebased this on saturday and that did fix the js errors I was getting. This time I have some errors on the network when loading the youtube transcripts. Have you seen this before?

chrisndodge commented:
run jenkins lettuce

felipemontoya commented:
@chrisndodge I ran the local equivalent, which would be `paver test_acceptance` and it did with no errors. Can I run the check again remotely on the jenkins server?

chrisndodge commented:
@felipemontoya yea, sometimes we have some "flakey" acceptance tests (either Lettuce or Bokchoy). In most cases there an async operation that the test doesn't properly wait for completion and there can be some timing related issues.

To re-run a test suite, in Jenkins, just write a comment:

jenkins run <suite name>

Sorry, looking up, I see I got the command wrong. Let me re-try

chrisndodge commented:
jenkins run lettuce

chrisndodge commented:
Hi @felipemontoya as next steps we should get some unit tests going for this...

felipemontoya commented:
I agree with that, unit testing this is necessary for it to used more widely.

chrisndodge commented:
@felipemontoya mind if I take over this PR? I'll probably close this PR and reopen it against a branch of mine. I think we need to add tests here, and then we can be good to go.

I want to do some prep work for the Hackathon and I should have some 'off hours' spare time to put into these microsite related improvements before then.

felipemontoya commented:
Hi @chrisndodge, I dont mind at all, if you want to take over this in a new branch. I have been meaning to do some more work here for the testing, but I have been unable to allocate some time to it.

chrisndodge commented:
@felipemontoya thanks.

Actually, can you close this PR and open a new one on cdodge/microsite-improvements?
